### ISSUE TITLE
pipeline documentation out dated / inconsistent

### ISSUE THREAD

**Issue by ewenmcneill** [faucetsdn/faucet#2797]
The [documentation of the Faucet Pipeline](https://docs.faucet.nz/en/latest/architecture.html#faucet-openflow-switch-pipeline) has:

*   a diagram with 9 tables (0 through 8)

*   a description of 10 tables (0 through 9)

which itself isn't consistent.  I *think* the one missing from the diagram is [`eth_dst_hairpin`](https://docs.faucet.nz/en/latest/architecture.html#table-7-eth-dst-hairpin) (table 7).

However the [*actual code*](https://github.com/faucetsdn/faucet/blob/master/faucet/faucet_pipeline.py) appears to have tables 0 through 11, with table 3 (`classification`) apparently not used by default:

```
ewen@faucet:~/tutorials$ cat list_faucet_tables 
#! /usr/bin/python3
import faucet.faucet_pipeline
table_mapping = {v.table_id: v.name for v in faucet.faucet_pipeline.FAUCET_PIPELINE}

for k in sorted(table_mapping.keys()):
    print("{0:2d}: {1}".format(k, table_mapping[k]))
ewen@faucet:~/tutorials$ ./list_faucet_tables 
 0: port_acl
 1: vlan
 2: vlan_acl
 3: classification
 4: eth_src
 5: ipv4_fib
 6: ipv6_fib
 7: vip
 8: eth_dst_hairpin
 9: eth_dst
10: flood
11: egress
ewen@faucet:~/tutorials$ 
```

which is even more tables.

For further confusion with the [NFV Tutorial](https://docs.faucet.nz/en/latest/tutorials/nfv_services.html) loaded it appears tables 0 through 7 are used (8 tables total), although in practice table 7 is empty and table 5 is basically empty (just "goto table 6"):

```
ewen@faucet:~/tutorials$ sudo ovs-ofctl -O OpenFlow13 dump-tables br0 | grep -v ditto
OFPST_TABLE reply (OF1.3) (xid=0x2):
  table 0:
    active=8, lookup=2751, matched=2751

  table 1:
    active=3, lookup=2736, matched=2736

  table 2:
    active=19, lookup=2678, matched=2678

  table 3:
    active=11, lookup=43, matched=43

  table 4:
    active=12, lookup=2465, matched=2465

  table 5:
    active=1, lookup=440, matched=440

  table 6:
    active=19, lookup=334, matched=334

  table 7:
    active=0, lookup=0, matched=0

ewen@faucet:~/tutorials$ 
```

This makes it difficult to match up the OpenFlow tables actually defined on the switch against their intended roles / usage in Faucet.

Does Faucet explicitly use the `table_id` values of the `faucet_pipeline.py` tables as the OpenFlow tables?  Or does it simply define tables onto the switch sequentially ignoring the tables it is not using.  If the `table_id` in `faucet_pipeline.py` is not actually used as the OpenFlow table ID, is there some way to map back from the OpenFlow table ID to what Faucet is trying to use it for to help with top to bottom understanding of how Faucet is working?

Ewen

**Comment by KitL** [comment]
The diagram is out of date. We have tried to automate building it, but we
have issues with including dependencies in the CI. It's also a little bit
misleading because the tables that actually get used depend on your
configuration. For instance the classification table is an option for when
you have a lot of tables available, it just has flows that would otherwise
be in the eth src table. I think our pipeline has reached the stage where
representing it in a single diagram isnt really very practical any more.

And unfortunately, the answer to most of your questions is it depends.

So there are potentially up to 11 tables, depending on the configuration. I
am working on adding a 12th.

In general these will be numbered sequentially, but that depends on the
hardware.

I think the best way to expose the mapping is probably to publish the
mapping between table names and table ids in prometheus. I think this info
is already available if you extract it from other fields but it makes sense
to make this an explicit variable. It probably also makes sense to log the
mapping as well.

Cheers,

Chris

**Comment by ewenmcneill** [comment]
[`blockdiag`](http://blockdiag.com/en/blockdiag/examples.html) seems to get pretty close to being able to generate a useful diagram (the main issue seems to be that it wants to draw some unrelated lines on top of each other, which is a bit confusing).  Conveniently it's implemented in Python, can be [embedded in Sphinx documentation](http://blockdiag.com/en/blockdiag/embed_diagram.html#embed-in-sphinx), and [integrated with Read The Docs](http://blockdiag.com/en/blockdiag/embed_diagram.html#embed-in-read-the-docs).

From a quick play with it, this input:

```
blockdiag {
  orientation = portrait
  port_acl [ label = "0: port_acl (M)" ]
  vlan     [ label = "1: vlan"         ]
  vlan_acl [ label = "2: vlan_acl (M)" ]
  eth_src  [ label = "3: eth_src (C)"  ]
  ipv4_fib [ label = "4: ipv4_fib"     ]
  ipv6_fib [ label = "5: ipv6_fib"     ]
  vip      [ label = "6: vip (C)"      ]
  eth_dst  [ label = "7: eth_dst (O)"  ]
  flood    [ label = "8: flood (O)"    ]

  port_acl -> vlan;
  vlan     -> vlan_acl;
  vlan     -> eth_src;
  vlan_acl -> eth_src;
  eth_src  -> vip;
  eth_src  -> ipv4_fib;
  eth_src  -> ipv6_fib;
  eth_src  -> eth_dst;
  ipv4_fib -> vip;
  ipv4_fib -> eth_dst;
  ipv6_fib -> vip;
  ipv6_fib -> eth_dst;
  vip      -> eth_dst;
  vip      -> flood;
  eth_dst  -> flood;
}
```

gets *fairly* close to the diagram in the existing documentation (see attached `.png` output), but for some overlapping edges creating confusion (which seems to be a `blockdiag` layout engine weakness that perhaps could be improved).

That input was hand written.  But it seems to be relatively easy to generate the required input from the `faucet_pipeline.py` code; see the attached quick hack (`gendiag.py`), and its output.

Ewen

![Faucet Existing Diagram](https://user-images.githubusercontent.com/4283973/52771800-bec29c80-309b-11e9-93d7-04de8cab7d3e.png)
[faucet.diag](https://github.com/faucetsdn/faucet/files/2863948/faucet-diag.txt)
[gendiag.py](https://github.com/faucetsdn/faucet/files/2863947/gendiag-py.txt)
![Faucet Code Diagram](https://user-images.githubusercontent.com/4283973/52773017-dea78f80-309e-11e9-855c-997e2839d8ef.png)

**Comment by ewenmcneill** [comment]
@KitL Thanks for all the comments.

Separate from the diagram (which as per my previous comment I think is solvable with a bit more layout smarts in something like `blockdiag` to avoid overlaying unrelated edges on the same pixels), there's the more fundamental problem of helping the user to map the Faucet stages they're expecting onto the OpenFlow tables that actually implement them.

I've got basically two use cases here:

*    I want to be able to run some command which "shows me the OpenFlow contents of the `eth_dst` table" (or some other arbitrary table by symbol name); and

*    When there is an OpenFlow action like "goto_table:5", I'd like to programmatically translate that into something which reveals which Faucet stage  is reached as a result (eg, "goto eth_dst")

I had originally hoped that it'd be possible to extract `table_id` and `name` from `faucet_pipeline.py` for each stage, and use that as a mapping, but:

*    The `table_id` values in the code do not (currently, consistently) match the text documentation *or* the diagram; and

*   (more importantly) the `table_id` values in `faucet_pipeline.py` do not match the OpenFlow table IDs either (because "certain tables may be omitted (**adjusting table IDs accordingly**) if the functionality is not required" -- emphasis added)

Having discovered the second point, I'm unclear what purpose the carefully calculated `table_id` figures in `faucet_pipeline.py` actually serve -- ie, whether they relate to *anything* outside the specific object in which they're defined.  (The inter-table linkages in `faucet_pipeline.py` appear to be by table *name*.)

FWIW, last time I dealt with this problem I maintained a 1:1 mapping from the Python code `table_id` tables to the OpenFlow table numbers.  If certain tables were not used, then they were just (a) empty and (b) never reached.  Providing the specific implementation allowed more tables than the highest possible `table_id` in the Python code, this didn't seem to cause any practical problems, and made user observability *much* easier (since there was a fixed mapping everything else could rely on, for a given version of the code).  (I guess there are OpenFlow 1.3 switches that are otherwise suitable that have fewer than 12 possible tables, so maybe the "renumber to eliminate omitted table numbers" behaviour might need to be retained as a backup option; but my experience of trying *many* switches a few years ago was that the ones that only supported small numbers of tables tended to *also* be *very* opinionated about what you could do in each table, typically exactly matching ASIC stages of a chipset originally designed for something else.)

The other reasonable approach is to expose tooling, or the raw data (via a REST API, returning JSON perhaps) that allows the *active* translation of "OpenFlow Table Number `<->` Faucet table role" to be retrieved at runtime (and thus mixed into tool wrappers, such as around `ovs-ofctl`).

I guess logging the resulting mapping to the `faucet.log` file, or sending it to monitoring tools (like Promethesus) is better than having no observability.... but the necessary information is still not easily retrieved by any other tooling.  So I'd suggest exposing it via a REST API or similar, in a machine parseable format, if possible.

Ewen

**Comment by KitL** [comment]
We have a tool for this purpose---fctl
<https://github.com/faucetsdn/faucet/blob/master/faucet/fctl.py>, though it
isnt really well documented, and is very bare bones. And as it turns out,
we already do publish the mapping between table names and id numbers to
prometheus, I just missed it earlier.

I was writing you an example of how you could extend this tool to do what
you want, but then I realised that if I am writing the code I should push
it back, so its in this PR <https://github.com/faucetsdn/faucet/pull/2798>.

So with this pr you could use something like:

import faucet.fctl as fctl
samples = fctl.get_samples('http://127.0.0.1:9302',
'faucet_config_table_names', {'dp_name': 'br1', 'table_name': 'eth_src'})
eth_src_table_id = samples[0].value

It would be good to make this a lot more usable. Probably include a
function like get_config_table_names for all the prometheus variables we
expose. Maybe there is a clever way to generate those functions
automatically. I'll write this stuff up in an issue (or you could) some
time.

The default table ids are calculated because some vendors do require that
we keep the table numbering consistent. And other vendors cant support 12
tables, so we need both options. I think its a reasonable argument that we
should try to keep the numbering consistent by default.

It would be good to have fctl be able to return the full flow table as
well, and when it does so it could rename goto actions. But we dont really
have a good way of exposing the flow table externally at the moment.

Cheers,

Chris

**Comment by ewenmcneill** [comment]
Thanks for the suggestion.  Definitely pushing code that makes it easier to explain how to do something is a good idea :-)

After debugging that example (see below), I've managed to extract something that looks both plausible and useful:

```
(venv) ewen@faucet:~/tutorials$ cat ./get-faucet-table-ids 
#! /usr/bin/env python3
#
# Based on example by @KitL from: https://github.com/faucetsdn/faucet/issues/2797

import faucet.fctl as fctl 

samples = fctl.get_samples(['http://127.0.0.1:9302' ],
                           'faucet_config_table_names', 
                           {'dp_name': 'sw1' })

table_id_to_name = { 
    sample.value: sample.labels['table_name']
    for sample in samples
}
for table_id in sorted(table_id_to_name.keys()):
    print("{0:2.0f}: {1}".format(table_id, table_id_to_name[table_id]))
(venv) ewen@faucet:~/tutorials$ ./get-faucet-table-ids 
 0: vlan
 1: vlan_acl
 2: eth_src
 3: ipv4_fib
 4: vip
 5: eth_dst
 6: flood
(venv) ewen@faucet:~/tutorials$ 
```

(and different from all the documentation, due to the renumbering of tables, so I think probably including `table_id` values in the documentation is counter productive to user understanding, not to mention likely to get out of date more quickly even if a direct mapping of faucet to OpenFlow table IDs is maintained).

Thanks for that.

There's... some rough edges for potential users including:

*    `get_samples()` expects an *interable* of endpoints to query, not a string as in the example given (resulting in a confusing `unknown url type: 'h'`)

*    the `dp_name` filter specified needs to be the *faucet* name of the data plane not the Open vSwitch name of the dataplane (and so far I don't know where *that* association can be found, other than by "there's only one" / guessing), so it needs to be, eg, `sw1` rather than `br1` or `br0`, etc.

*    the returned table ID values are *floats* not integers, which I assume is a surprising side effect resulting from this scraping a Promethesus monitoring API rather than an actual API designed for retrieving this information for direct use

But at least the information can be obtained, and seems valid.

Some more random notes below in case they're of help to someone else trying to do this.

Ewen

When attempting to use this via the suggested example from @KitL in https://github.com/faucetsdn/faucet/issues/2797#issuecomment-463889726, it ends up getting confused somewhere around https://github.com/faucetsdn/faucet/blob/f3c28cef1100e76ae32eb70c28b0219a5adad61c/faucet/fctl.py#L51 by the single endpoint being specified, apparently due expecting an interable of endpoints, not a single one.

```
(venv) ewen@faucet:~/tutorials$ cat get-faucet-table-ids 
#! /usr/bin/env python3
#
# Based on example by @KitL from: https://github.com/faucetsdn/faucet/issues/2797

import faucet.fctl as fctl 

samples = fctl.get_samples('http://127.0.0.1:9302', 
                           'faucet_config_table_names', 
                           {'dp_name': 'br1', 'table_name': 'eth_src'}) 
eth_src_table_id = samples[0].value
(venv) ewen@faucet:~/tutorials$ ./get-faucet-table-ids 
unknown url type: 'h'Traceback (most recent call last):
  File "./get-faucet-table-ids", line 10, in <module>
    eth_src_table_id = samples[0].value
TypeError: 'NoneType' object is not subscriptable
(venv) ewen@faucet:~/tutorials$ 
```

which seems like something that maybe should be detected (ie, the "iterable" that was passed in is actually a str / bytearray, and should be treated as a singleton and/or rejected).

The `dp_name` needed to be the *faucet* name of the dataplane not the *Open vSwitch name* (so `sw1` not `br1` or `br0`) -- which I found by using `None` as the label match criteria to get a wider dump -- I think I have something working.  (IMHO `None` should be the default argument for the label match criteria, or at least documented as a reasonable option; I found it only by reading the logic of `_get_samples_from_metrics`...)

```
(venv) ewen@faucet:~/tutorials$ cat test-get-samples 
#! /usr/bin/env python3
#
# Based on example by @KitL from: https://github.com/faucetsdn/faucet/issues/2797

import faucet.fctl as fctl 

samples = fctl.get_samples(['http://127.0.0.1:9302' ],
                           'faucet_config_table_names', 
                           None)
print(samples)
(venv) ewen@faucet:~/tutorials$ ./test-get-samples 
[Sample(name='faucet_config_table_names', labels={'dp_id': '0x1', 'dp_name': 'sw1', 'table_name': 'eth_dst'}, value=5.0, timestamp=None, exemplar=None), Sample(name='faucet_config_table_names', labels={'dp_id': '0x1', 'dp_name': 'sw1', 'table_name': 'eth_src'}, value=2.0, timestamp=None, exemplar=None), Sample(name='faucet_config_table_names', labels={'dp_id': '0x1', 'dp_name': 'sw1', 'table_name': 'flood'}, value=6.0, timestamp=None, exemplar=None), Sample(name='faucet_config_table_names', labels={'dp_id': '0x1', 'dp_name': 'sw1', 'table_name': 'ipv4_fib'}, value=3.0, timestamp=None, exemplar=None), Sample(name='faucet_config_table_names', labels={'dp_id': '0x1', 'dp_name': 'sw1', 'table_name': 'vip'}, value=4.0, timestamp=None, exemplar=None), Sample(name='faucet_config_table_names', labels={'dp_id': '0x1', 'dp_name': 'sw1', 'table_name': 'vlan'}, value=0.0, timestamp=None, exemplar=None), Sample(name='faucet_config_table_names', labels={'dp_id': '0x1', 'dp_name': 'sw1', 'table_name': 'vlan_acl'}, value=1.0, timestamp=None, exemplar=None)]
(venv) ewen@faucet:~/tutorials$ 
```

If anyone else wants to try to reproduce this, I [checked out the change from this pull request](https://help.github.com/articles/checking-out-pull-requests-locally/#modifying-an-inactive-pull-request-locally) with:

```
(venv) ewen@faucet:~/faucet$ git checkout master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
(venv) ewen@faucet:~/faucet$ git fetch origin pull/2798/head:fcntl-2798
remote: Enumerating objects: 8, done.
remote: Counting objects: 100% (8/8), done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 8 (delta 5), reused 8 (delta 5), pack-reused 0
Unpacking objects: 100% (8/8), done.
From https://github.com/faucetsdn/faucet
 * [new ref]           refs/pull/2798/head -> fcntl-2798
(venv) ewen@faucet:~/faucet$ git checkout fcntl-2798
Switched to branch 'fcntl-2798'
(venv) ewen@faucet:~/faucet$
(venv) ewen@faucet:~/faucet$ pip install --editable .
```

(also beware that if you have a `#! /usr/bin/python3` line at the start of your script, you'll defeat the `venv` handling, and thus load from the wrong Python path, which will probably load a system-wide installed version, which doesn't have these changes.  Ask me how I know this :-) )

**Comment by ewenmcneill** [comment]
FWIW, Open vSwitch is perfectly happy to create a bridge called `sw1` if you want, using `ovs-vsctl`.  But it seems like Faucet is defaulting to using/calling it `br0`, even when the `/etc/faucet/faucet.yaml` file gives the `dps` another name.  (I'm unclear if this is a result of Open vSwitch default bridges and `dp_id` matching, or something else faucet is doing.)

Possibly something within Ryu (eg https://github.com/osrg/ryu/blob/master/ryu/lib/ovs/bridge.py or https://github.com/osrg/ryu/blob/master/ryu/lib/ovs/vsctl.py), since it looks like the Faucet code doesn't invoke `ovs-vsctl` directly itself.  It looks like Open vSwitch [does not support renaming a bridge after creatiion](https://mail.openvswitch.org/pipermail/ovs-discuss/2012-March/026502.html).

(If Faucet does pass down the bridge name via Ryu to Open vSwitch, then it might be worth *not* using the default `br0`'s `dp_id` in the tutorial examples, so that the dataplanes used by Faucet have the same names in Faucet and Open vSwitch....  although it looks like the *default* `ovs-ofctl` behaviour on creating a new bridge is to give it a *random* `dp_id`, so maybe the default bridge is *not* the issue here.)

Anyway, for *fetching* table information best workaround might be to get the `dp_id` off the Open vSwitch with, eg,

```
ewen@faucet:~/faucet$ printf "0x%x\n" $(sudo ovs-ofctl show br0 | awk -F : '/dpid:/ {print $3;}')
0x1
ewen@faucet:~/faucet$ 
```

(converting it from the 0-padded `ovs-ofctl` output to the `0xN` format the Faucet expects, because the `dp_id` value that comes back for matching from the Promethesus interface is a string not an integer, sigh).

And then use that for matching on the required dataplane / table entries:

```
(venv) ewen@faucet:~/tutorials$ cat ./get-faucet-table-ids 
#! /usr/bin/env python3
#
# Example by @KitL from: https://github.com/faucetsdn/faucet/issues/2797

import faucet.fctl as fctl 

samples = fctl.get_samples(['http://127.0.0.1:9302' ],
                           'faucet_config_table_names', 
                           {'dp_id': '0x1' })

table_id_to_name = { 
    sample.value: sample.labels['table_name']
    for sample in samples
}
for table_id in sorted(table_id_to_name.keys()):
    print("{0:2.0f}: {1}".format(table_id, table_id_to_name[table_id]))
(venv) ewen@faucet:~/tutorials$ ./get-faucet-table-ids 
 0: vlan
 1: vlan_acl
 2: eth_src
 3: ipv4_fib
 4: vip
 5: eth_dst
 6: flood
(venv) ewen@faucet:~/tutorials$ 
```

Ewen

```
ewen@faucet:~$ sudo ovs-vsctl list-br
br0
ewen@faucet:~$ sudo ovs-vsctl add-br sw1
ewen@faucet:~$ sudo ovs-vsctl list-br
br0
sw1
ewen@faucet:~$ 
ewen@faucet:~$ cat /etc/faucet/faucet.yaml
acls:
    mirror-acl:
        - rule:
            actions:
                allow: true
                mirror: zeek
vlans:
    zeek-vlan:
        vid: 100
        description: "Zeek IDS network"
    vlan200:
        vid: 200
        description: "192.168.2.0/24 network"
        faucet_vips: ["192.168.2.254/24"]
        faucet_mac: "00:00:00:00:00:22"
        acls_in: [mirror-acl]
    vlan300:
        vid: 300
        description: "192.168.3.0/24 network"
        faucet_vips: ["192.168.3.254/24"]
        faucet_mac: "00:00:00:00:00:33"
        acls_in: [mirror-acl]
dps:
    sw1:
        dp_id: 0x1
        hardware: "Open vSwitch"
        interfaces:
            1:
                name: "host1"
                description: "host1 network namespace"
                native_vlan: vlan200
            2:
                name: "host2"
                description: "host2 network namespace"
                native_vlan: vlan200
            3:
                name: "host3"
                description: "host3 network namespace"
                native_vlan: vlan300
            4:
                name: "host4"
                description: "host4 network namespace"
                native_vlan: vlan300
            5:
                name: "dnsmasq"
                description: "dnsmasq server network namespace"
                tagged_vlans: [vlan200, vlan300]
            6:
                name: "zeek"
                description: "Zeek network namespace"
                native_vlan: zeek-vlan
ewen@faucet:~$ 
```

```
ewen@faucet:~/faucet$ sudo ovs-ofctl show br0
OFPT_FEATURES_REPLY (xid=0x2): dpid:0000000000000001
n_tables:254, n_buffers:0
capabilities: FLOW_STATS TABLE_STATS PORT_STATS QUEUE_STATS ARP_MATCH_IP
actions: output enqueue set_vlan_vid set_vlan_pcp strip_vlan mod_dl_src mod_dl_dst mod_nw_src mod_nw_dst mod_nw_tos mod_tp_src mod_tp_dst
 LOCAL(br0): addr:be:a7:70:48:59:4b
     config:     PORT_DOWN
     state:      LINK_DOWN
     speed: 0 Mbps now, 0 Mbps max
OFPT_GET_CONFIG_REPLY (xid=0x4): frags=normal miss_send_len=0
ewen@faucet:~/faucet$ 
ewen@faucet:~/faucet$ sudo ovs-ofctl show sw1
OFPT_FEATURES_REPLY (xid=0x2): dpid:00005ac015c71a43
n_tables:254, n_buffers:0
capabilities: FLOW_STATS TABLE_STATS PORT_STATS QUEUE_STATS ARP_MATCH_IP
actions: output enqueue set_vlan_vid set_vlan_pcp strip_vlan mod_dl_src mod_dl_dst mod_nw_src mod_nw_dst mod_nw_tos mod_tp_src mod_tp_dst
 LOCAL(sw1): addr:5a:c0:15:c7:1a:43
     config:     PORT_DOWN
     state:      LINK_DOWN
     speed: 0 Mbps now, 0 Mbps max
OFPT_GET_CONFIG_REPLY (xid=0x4): frags=normal miss_send_len=0
ewen@faucet:~/faucet$ 
```

**Comment by byllyfish** [comment]
I agree that the table_id's of the faucet pipeline don't need to appear in the documentation.

One thing that might help is to make the 'faucet_config_table_names' info metric include the next_tables as a label of comma-separated values, e.g. `next_tables="vlan,eth_src"`. With this change, the running pipeline could be visualized using the fctl tool.

**Comment by ewenmcneill** [comment]
@byllyfish Having a `next_tables="..."` label seemed like a useful idea to me too.  So I've made an attempt to implement it, in https://github.com/faucetsdn/faucet/pull/2804, which seems to produce plausibly useful results:

```
(venv) ewen@faucet:~/tutorials$ cat get-faucet-table-ids 
#! /usr/bin/env python3
#
# Example by @KitL from: https://github.com/faucetsdn/faucet/issues/2797

import faucet.fctl as fctl 

samples = fctl.get_samples(['http://127.0.0.1:9302' ],
                           'faucet_config_table_names', 
                           {'dp_id': '0x1' })

#print(samples)
table_id_to_name = { 
    sample.value: sample.labels['table_name']
    for sample in samples
}
table_name_to_next_tables = { 
    sample.labels['table_name']: sample.labels['next_tables'].split(',')
    for sample in samples
}
for table_id in sorted(table_id_to_name.keys()):
    table_name = table_id_to_name[table_id]
    next_tables = table_name_to_next_tables[table_name]

    print("{0:2.0f}: {1:20s} {2}".format(table_id, table_name, next_tables))
(venv) ewen@faucet:~/tutorials$ ./get-faucet-table-ids 
 0: port_acl             ['vlan', 'vip', 'eth_dst', 'flood']
 1: vlan                 ['vlan_acl', 'eth_src']
 2: vlan_acl             ['eth_src', 'eth_dst', 'flood']
 3: eth_src              ['ipv4_fib', 'ipv6_fib', 'vip', 'eth_dst', 'flood']
 4: ipv4_fib             ['vip', 'eth_dst', 'flood']
 5: ipv6_fib             ['vip', 'eth_dst', 'flood']
 6: vip                  ['eth_dst', 'flood']
 7: eth_dst              ['']
 8: flood                ['']
(venv) ewen@faucet:~/tutorials$ 
```

**Comment by byllyfish** [comment]
This looks pretty good. I think there is a missing transition from the eth_dst table to the flood table. The table config also has a miss_goto value, which is separate from next_tables (it turns into next_tables_miss in the TableFeaturesMessage). The eth_dst table's `miss_goto` goes to the flood table.

`valve_table.table_config.miss_goto` is the name of the table.

In some cases, the miss_goto table is already included in next_tables.

**Comment by ewenmcneill** [comment]
@byllyfish wrote:
> This looks pretty good. I think there is a missing transition from the eth_dst table to the flood table. 

So there is.  Overlooking `miss_goto` is indeed why it doesn't look right.  Thanks for pointing that out.  I've updated https://github.com/faucetsdn/faucet/pull/2804 to include `miss_goto` out of the `table_config` if set, and not already present.

As noted in the pull request update, the new output (for the facuet git example config) is:

```
(venv) ewen@faucet:~/tutorials$ ./get-faucet-table-ids 
 0: port_acl             ['vlan', 'vip', 'eth_dst', 'flood']
 1: vlan                 ['vlan_acl', 'eth_src']
 2: vlan_acl             ['eth_src', 'eth_dst', 'flood']
 3: eth_src              ['ipv4_fib', 'ipv6_fib', 'vip', 'eth_dst', 'flood']
 4: ipv4_fib             ['vip', 'eth_dst', 'flood']
 5: ipv6_fib             ['vip', 'eth_dst', 'flood']
 6: vip                  ['eth_dst', 'flood']
 7: eth_dst              ['flood']
 8: flood                ['']
(venv) ewen@faucet:~/tutorials$ 
```

And there's probably now enough output that the `blockdiag` config generator I hacked together in https://github.com/faucetsdn/faucet/issues/2797#issuecomment-463536734 could be adapted to work via `fcntl` instead of parsing the code structures directly.  If https://github.com/faucetsdn/faucet/pull/2804 gets merged, I'll have a look at doing that sometime.

Ewen

**Comment by anarkiwi** [comment]
Checking in - looks like original problem (stale table ID references, among others has been addressed). Is the proposed next stage to extend fctl to dump the current pipeline in a convenient format?

**Comment by ewenmcneill** [comment]
@anarkiwi Yes, the documentation is less confusing now (doesn't reference OpenFlow table IDs which are probably not being used in practice). 

In addition `fctl` and the `faucet_config_table_names` now should allow retrieving enough information at runtime to determine the *actual* OpenFlow pipeline that is active, and the mapping of active OpenFlow Table IDs to Faucet pipeline stages.  And thus the user cross-referencing against the output of `ovf-ofctl`, which was the problem I first started with.

There are a few potential follow up items:

*    Dynamically generate graph of "possible full pipeline" and/or "current running pipeline" (rather than hand maintained one which is known to be out of date / not match what is active on user's system)

*   Documentation is possibly still not complete list of all the possible tables used (apparently some were added since 2017)

*   `ovf-ofctl` wrapper (or equivalent tool) that output table IDs as Faucet table names (instead of or in addition to the OpenFlow Table IDs) to assist user with understanding active rules

But all of those could be addressed separately.

So if y'all would like to close this issue now, that seems okay with me.

Ewen

PS: Yes at some point I do intend to look at updating the `gendiag.py` script (attached to https://github.com/faucetsdn/faucet/issues/2797#issuecomment-463536734) to use the `fctl` pipeline output rather than the digging around in the Faucet classes for what turns out to be a superset of possible Faucet pipelines.  But that too can be done without this issue staying open (it's on my work to do list separately).
